instaclustr/resource_kafka_topic.go

```diff
@@ -0,0 +1,382 @@
+	}
+
+	// It won't read the topic's config if the config{} block is missing in the resource
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] A block of config{} needs to be provided in the resource to read the topic's config.")
+		log.Printf("[INFO] Topic %s's config is not read.", topic)
+		return nil
+	}
+
+	config := make(map[string]interface{}, 0)
+	configList := make([]map[string]interface{}, 0)
+	err = mapstructure.Decode(kafkaTopicConfig.Config, &config)
+	if err != nil {
+		return err
+	}
+	configList = append(configList, config)
+	err = d.Set("config", configList)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Successfully fetch the config of topic %s.", topic)
+	return nil
+}
+
+// There are two use cases of resourceKafkaTopicUpdate:
+// 1. Providing all the possible configs in resource, and modify the ones you want to update. Then keeping all the
+// configs in the resource.
+// 2. Providing some configs (not all) you want to update, and then keep only these configs in the resource.
+// Terraform always detects that there is an update in case 2 if you run the plan again, because resourceKafkaTopicRead
+// reads all the configs, but there are only some of them provided in resource. For the missing configs, Terraform just
+// sets them as their default values.
+//
+// Although Terraform detects an update, func getChangedConfigMap is used to handle case 2. It makes sure that there
+// will not be a real update if the configs in the resources are not changed. So, it's safe to keep only partial of
+// configs in the config{} block, there won't be an update request if their values are same as what have been read.
+func resourceKafkaTopicUpdate(d *schema.ResourceData, meta interface{}) error {
+	cluster_id := d.Get("cluster_id").(string)
+	topic := d.Get("topic").(string)
+	log.Printf("[INFO] Updating the config of Kafka topic %s in %s.", topic, cluster_id)
+
+	// This is for when only replication_factor or partitions are changed, but we don't support changing them currently
+	if !d.HasChange("config") {
+		log.Printf("[INFO] Currently we only supprot updating topic's config. There are no changes in topic %s's config.", topic)
+		return nil
+	}
+	// This is for when there is no config{} block in the resource, or the whole config{} is removed from the resource.
+	if len(d.Get("config").([]interface{})) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		return nil
+	}
+	client := meta.(*Config).Client
+
+	changedConfigMap, err := getChangedConfigMap(d)
+	if err != nil {
+		return err
+	}
+	if len(changedConfigMap) == 0 {
+		log.Printf("[INFO] There are no configs to be updated.")
+		err = resourceKafkaTopicRead(d, meta)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	var kafkaTopicConfigOptions KafkaTopicConfigOptions
+	err = mapstructure.Decode(changedConfigMap, &kafkaTopicConfigOptions)
+	if err != nil {
```

Comment:
I read somewhere or other that the preferred way of propagating errors in go is to chain them with context. I'm not a go expert so I'm not sure if that true or not but here is an example anyway:

<pre>
func (thing *Thingo) Thingo() error {
    if err := thing.DoSomething(); err != nil {
        return fmt.Errorf("could not do the thing: %w", err)
    }
    ...
}
</pre>

Reply:
Yeah, you're right, will add the context.

Reply:
if you are following this method, I believe you should use the %w specifier with 'err' operands

Reply:
Ah, forgot about this, will add it.